### PR TITLE
Normalize l2

### PR DESCRIPTION
**Pull request by amitibo**
This PR is only the l2 normalization layer from #1225.

### CODE REVIEW

File: chainer/functions/normalization/l2_normalization.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+        return gx,
+
+
+def normalize(x, eps=1e-5):
+    """L2 norm (a.k.a. Euclidean norm) squared.
+
+    This function implements L2 normalization on a 1D vector. No reduction
+    along batch axis is done.  Let :math:`x` be an input vector of dimension
+    :math:`(N,D)`, where :math:`N` and :math:`K` denote mini-batchsize and the
```
Comment (zori) [inline chainer/functions/normalization/l2_normalization.py]: (N, K)

File: chainer/functions/normalization/l2_normalization.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    def __init__(self, eps=1e-5):
+        self.eps = eps
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        x_type, = in_types
+
+        type_check.expect(
+            x_type.dtype == numpy.float32,
```
Comment (zori) [inline chainer/functions/normalization/l2_normalization.py]: x_type.dtype.kind == 'f',

File: chainer/functions/normalization/l2_normalization.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        return x / norm,
+
+    def backward(self, inputs, gy):
+        x = inputs[0]
+        gy = gy[0]
+
+        xp = cuda.get_array_module(x)
+
+        if xp is numpy:
+            norm = numpy.sqrt((x * x).sum(axis=1)) + self.eps
```
Comment (zori) [inline chainer/functions/normalization/l2_normalization.py]: `numpy.linalg.norm` here as well

File: chainer/functions/normalization/l2_normalization.py (hunk 2 above)
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: Supporting all floating types is not mandatory. It is OK only for float32.

File: chainer/functions/normalization/l2_normalization.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            'l2norm'
+        )
+        norm = cuda.cupy.broadcast_to(
+            l2norm_kernel(x, self.eps, axis=1).reshape(-1, 1),
+            x.shape
+        )
+
+        return x / norm,
+
+    def backward(self, inputs, gy):
```
Comment (zori) [inline chainer/functions/normalization/l2_normalization.py]: Please split into `backward_cpu` and `backward_gpu`.

**Comment by zori** [comment]
LGTM. Thank you.

**Pull request merged**